in-progress/proving-queue/0005-proving-queue.md

```diff
@@ -0,0 +1,97 @@
+
+![alt text](./broker.png)
+
+The entire proof request index DB is loaded into memory upon service start.
+
+Prover agents periodically poll the broker to either request more work or to update the broker that they are still working on a specific job. As there are a large number of clients we will need to limit the frequency of that polling to, perhaps every 10 seconds.
+
+This gives an overall job flow that looks as follows:
+
+1. The broker receives a `Proof Request` for job id #4567 from the orchestrator.
+2. The broker persists the proof's input data to disk and then inserts an entry into the index DB. Finally it creates an entry in the in-memory cache of jobs.
+3. An agent polls for new work. A query is performed on the in-memory cache based on the capabilites of the prover and ordered to prioritise earlier epochs, discounting all jobs that are already being proven.
+4. The query returns the job #4567. 
+5. The broker stores the prover agent id and the current time against the job as `Start Time`.
+6. The broker returns the job details to the agent.
+
+Later, the agent polls again:
+
+1. The agent polls the broker with the job id and job start time
+2. The broker finds the job in the in-memory cache
+3. If the job is being worked on by another agent then the start time against the job is compared to the start time in the agent's request. The earlier time wins and the cache is updated if necessary with the winner's details. If the polling agent loses then the agent is told to stop working on the job and is given a new job.
+4. The broker adds the current time to the job as `Last Update Time`
+
+Later, the agent polls to say the job is completed:
+
+1. The agent has completed the job and immediately polls with the result.
+2. The results are stored and the job is updated as completed, probably moved to a different in memory cache.
+
+The reason for the start time negotiation around agents is that we need to consider a broker crash/restart.
+
+Upon restart, the broker should:
+
+1. Re-build the in-memory cache from the index.
+2. Start accepting new job requests from agents.
+
+The above means that a job that is already being worked could be given to another agent after a restart. All of the data around current proving jobs is only in memory so is lost upon process termination. By negotiating based on start time, soon after restart the system will revert to how it was without too much duplicated effort. Of course, for small proofs, it is still possible that a proof could be completed twice. This seems an acceptable trade-off.
+
+Finally, the broker periodically checks all current jobs to see if their `Last Update Time` is beyond a certain window. If so, the details of the agent proving the job are removed, so they become available again.
```

Comment:
I'd say we remove this periodic process, and instead when an agent polls we search for jobs that are either not taken or that their last update time is beyond a threshold.